File: src/statesman/commands/InventoryCommand.java
package statesman.commands;

import java.util.Iterator;

import statesman.Content;
import statesman.Interpreter;
import statesman.InventoryItem;

public class InventoryCommand extends Command {

    public enum InventoryActionType {
        ADD, REMOVE, CLEAR, LIST
    }

    public static final String ID_INV = "inv";
    public static final String ID_INV_ADD = "add";
    public static final String ID_INV_RM = "rm";
    public static final String ID_INV_CLEAR = "clear";
    public static final String ID_INV_LIST = "list";

    private String _itemName;
    private InventoryActionType _action;

    private InventoryCommand() {
        _itemName = null;
        _action = null;
    }

    public InventoryCommand(InventoryActionType action, String itemName) {
        boolean nameRequired = (action == InventoryActionType.ADD
                || action == InventoryActionType.REMOVE);
        if (nameRequired && itemName.isBlank()) {
            throw new IllegalArgumentException(
                    "Inventory item name cannot be blank");
        }
        if (action == null) {
            throw new IllegalArgumentException(
                    "Invalid value was passed to the action parameter");
        }
        _itemName = itemName;
        _action = action;
    }

    private static Command _defaultInstance;

    public static Command getDefault() {
        if (_defaultInstance == null) {
            _defaultInstance = new InventoryCommand();
        }
        return _defaultInstance;
    }

    private InventoryItem getItem() {
        if (_itemName == null) {
            return null;
        }
        return Interpreter.getScene().getItems().get(_itemName);
    }

    @Override
    public Command fromText(String commandId, String[] arguments) {
        if (arguments.length < 2) {
            return null;
        }

        String action = arguments[1].trim();
        InventoryActionType actionType = null;
        String actionValue = null;
        if (arguments.length == 2) {
            if (action.equalsIgnoreCase(ID_INV_LIST)) {
                actionType = InventoryActionType.LIST;
            } else if (action.equalsIgnoreCase(ID_INV_CLEAR)) {
                actionType = InventoryActionType.CLEAR;
            }
        } else if (arguments.length == 3) {
            if (action.equalsIgnoreCase(ID_INV_ADD)) {
                actionType = InventoryActionType.ADD;
            } else if (action.equalsIgnoreCase(ID_INV_RM)) {
                actionType = InventoryActionType.REMOVE;
            }
            actionValue = arguments[2].trim();
        }

        if (actionType == null) {
            return null;
        }
        return new InventoryCommand(actionType, actionValue);
    }

    @Override
    public void execute() {
        InventoryItem item = getItem();
        switch (_action) {
        case ADD:
            if (item == null) {
                break;
            }
            if (Interpreter.getInventory().containsKey(item.getName())) {
                System.out.println(Content.getScript().findMessage("i_1"));
                break;
            }
            Interpreter.getInventory().put(item.getName(), item);
            break;
        case REMOVE:
            if (item == null) {
                break;
            }
            if (Interpreter.getInventory().containsKey(item.getName())) {
                Interpreter.getInventory().remove(_itemName);
                break;
            }
            System.out.println(Content.getScript().findMessage("i_2"));
            break;
        case LIST:
            int inventorySize = Interpreter.getInventory().size();
            if (inventorySize > 0) {
                if (inventorySize == 1) {
                    System.out.println(Content.getScript().findMessage("i_3"));
                } else {
                    System.out.printf(Content.getScript().findMessage("i_4"),
                            inventorySize);
                }
                Iterator<InventoryItem> iterator = Interpreter.getInventory()
                        .values().iterator();
                while (iterator.hasNext()) {
                    InventoryItem currentItem = iterator.next();
                    System.out.printf(Content.getScript().findMessage("i_5"),
                            currentItem.getName(),
                            currentItem.getDescription());
                }
            } else {
                System.out.println(Content.getScript().findMessage("i_6"));
            }
            break;
        case CLEAR:
            Interpreter.getInventory().clear();
            break;
        default:
            break;
        }
    }

}

File: src/Commands/InventoryCommand.cs
namespace Statesman.Commands
{
    public class InventoryCommand : Command
    {
        public enum InventoryActionType
        {
            Add,
            Remove,
            Clear,
            List
        }

        public const string kIdInventory = "inv";

        public const string kInventoryAdd = "add";
        public const string kInventoryRemove = "rm";
        public const string kInventoryClear = "clear";
        public const string kInventoryList = "list";

        public string ItemName { get; }
        private InventoryItem Item
        {
            get
            {
                if (ItemName == null)
                {
                    return null;
                }
                return Interpreter.Scene.Items[ItemName];
            }
        }
        public InventoryActionType ActionType { get; set; }

        private InventoryCommand(InventoryActionType action, string itemName)
        {
            bool nameRequired = action == InventoryActionType.Add
                    || action == InventoryActionType.Remove;
            if (nameRequired && string.IsNullOrWhiteSpace(itemName))
            {
                throw new ArgumentException("Inventory item name cannot be blank");
            }
            ItemName = itemName;
            ActionType = action;
        }

        public static Command FromText(string commandId, string[] arguments)
        {
            if (commandId != kIdInventory || arguments.Length < 2)
            {
                return null;
            }

            string actionString = arguments[1].Trim().ToLowerInvariant();
            InventoryActionType? actionType = null;
            string actionValue = "";
            if (arguments.Length == 2)
            {
                if (actionString.Equals(kInventoryList))
                {
                    actionType = InventoryActionType.List;
                }
                else if (actionString.Equals(kInventoryClear))
                {
                    actionType = InventoryActionType.Clear;
                }
            }
            else if (arguments.Length == 3)
            {
                if (actionString.Equals(kInventoryAdd))
                {
                    actionType = InventoryActionType.Add;
                }
                else if (actionString.Equals(kInventoryRemove))
                {
                    actionType = InventoryActionType.Remove;
                }
                actionValue = arguments[2].Trim();
            }

            if (actionType.HasValue)
            {
                return new InventoryCommand(actionType.Value, actionValue);
            }
            return null;
        }

        public override void Execute()
        {
            switch (ActionType)
            {
                case InventoryActionType.Add:
                    if (Item == null)
                    {
                        return;
                    }
                    if (Interpreter.Inventory.ContainsKey(Item.Name))
                    {
                        Console.WriteLine(Content.Script.FindMessage("i_1"));
                        break;
                    }
                    Interpreter.Inventory.Add(Item.Name, Item);
                    break;
                case InventoryActionType.Remove:
                    if (Item == null)
                    {
                        return;
                    }
                    if (Interpreter.Inventory.Remove(ItemName))
                    {
                        break;
                    }
                    Console.WriteLine(Content.Script.FindMessage("i_2"));
                    break;
                case InventoryActionType.List:
                    var realInventory = Interpreter.Inventory.Values.Where((item) => !item.IsSwitch);
                    int inventorySize = realInventory.Count();
                    if (inventorySize > 0)
                    {
                        if (inventorySize == 1)
                        {
                            Console.WriteLine(Content.Script.FindMessage("i_3"));
                        }
                        else
                        {
                            Console.Write(Content.Script.FindMessage("i_4"), inventorySize);
                        }
                        foreach (var inventoryItem in realInventory)
                        {
                            Console.Write(
                                    Content.Script.FindMessage("i_5"),
                                    inventoryItem.Name,
                                    inventoryItem.Description);
                        }
                        Console.WriteLine();
                    }
                    else
                    {
                        Console.WriteLine(Content.Script.FindMessage("i_6"));
                    }
                    break;
                case InventoryActionType.Clear:
                    Interpreter.Inventory.Clear();
                    break;
                default:
                    break;
            }
        }
    }
}
